sopel/coretasks.py

```diff
@@ -540,82 +538,72 @@ def _parse_modes(bot, args, clear=False):
         "Y": plugin.OPER,
     }
 
-    modes = {}
-    if not clear:
-        # Work on a copy for some thread safety
-        modes.update(channel.modes)
-
-    # Process modes
-    sign = ""
-    param_idx = 0
-    chanmodes = bot.isupport.CHANMODES
-    for char in modestring:
-        # Are we setting or unsetting
-        if char in "+-":
-            sign = char
-            continue
-
-        if char in chanmodes["A"]:
-            # Type A (beI, etc) have a nick or address param to add/remove
-            if char not in modes:
-                modes[char] = set()
-            if sign == "+":
-                modes[char].add(params[param_idx])
-            elif params[param_idx] in modes[char]:
-                modes[char].remove(params[param_idx])
-            param_idx += 1
-        elif char in chanmodes["B"]:
-            # Type B (k, etc) always have a param
-            if sign == "+":
-                modes[char] = params[param_idx]
-            elif char in modes:
-                modes.pop(char)
-            param_idx += 1
-        elif char in chanmodes["C"]:
-            # Type C (l, etc) have a param only when setting
-            if sign == "+":
-                modes[char] = params[param_idx]
-                param_idx += 1
-            elif char in modes:
-                modes.pop(char)
-        elif char in chanmodes["D"]:
-            # Type D (aciLmMnOpqrRst, etc) have no params
-            if sign == "+":
-                modes[char] = True
-            elif char in modes:
-                modes.pop(char)
-        elif char in mapping and (
-            "PREFIX" not in bot.isupport or char in bot.isupport.PREFIX
-        ):
-            # User privs modes, always have a param
-            nick = Identifier(params[param_idx])
-            priv = channel.privileges.get(nick, 0)
-            value = mapping.get(char)
-            if value is not None:
-                if sign == "+":
-                    priv = priv | value
-                else:
-                    priv = priv & ~value
-                channel.privileges[nick] = priv
-            param_idx += 1
+    privileges = mapping.keys()
+    if 'PREFIX' in bot.isupport:
+        privileges = bot.isupport.PREFIX.keys()
+
+    modemessage = ircmodes.ModeMessage(
+        bot.isupport.CHANMODES,
+        privileges=set(privileges),
+    )
+    modeinfo = modemessage.parse_modestring(args[1], tuple(args[2:]))
```

Comment:
This, too, is overhead of which I'm wary—constructing a fresh object every time the bot gets a `MODE` line.

The parameters to this constructor won't change once 005 `RPL_ISUPPORT` lines end, so I think it would be better for the bot to contain its own mode parser object that is rebuilt only when receiving lines that might change its behavior. Maybe you don't want to do it as part of this PR, but that should be a refactor goal.

Reply:
Oh yeah, that could totally be an attribute of the bot, indeed.